server/expression.js

```diff
@@ -204,72 +233,75 @@ expression.getTFAmatLine_ = function(file, name) {
 };
 
 /**
- * Gets the expression matrix profile of a given gene.
+ * Gets the expression matrix profile of given genes and conditions.
  * @param {string} fileExp File name of the expression matrix.
  * @param {string} fileTFA File name of the TFA matrix.
- * @param {string} name Name of the gene to be profiled.
+ * @param {!Array<string>} geneNames Names of the expression matrix.
+ * @param {!Array<string>} conditionNames Names of the expression matrix.
  * @return {?expression.Profile} Gene expression profile as a JS object.
  * @private
  */
-expression.getExpmatLine_ = function(fileExp, fileTFA, name) {
+expression.getExpmatLine_ = function(fileExp, fileTFA, geneNames,
+                                     conditionNames) {
   var bufExp = utils.readFileToBuf(fileExp);
   var bufTFA = null;
   if (fileTFA != null) {
     bufTFA = utils.readFileToBuf(fileTFA);
   }
-
   if (bufExp == null) {
-    console.error('cannot read file', fileExp);
-    return null;
+    return console.error('cannot read file', fileExp), [];
   }
   if (fileTFA != null && bufTFA == null) {
-    console.error('cannot read file', fileTFA);
-    return null;
+    return console.error('cannot read file', fileTFA), [];
   }
 
   var resultExp = expression.readExpmat_(bufExp);
   var resultTFA;
   if (fileTFA != null) {
-    resultTFA = expression.readTFAmat_(/** @type {!Buffer} */(bufTFA));
+    resultTFA = expression.readTFAmat_(/** @type {!Buffer} */bufTFA);
   }
 
-  for (var i = 0; i < resultExp.rownames.length; i++) {
-    if (resultExp.rownames[i].toLowerCase() == name) {
-      name = resultExp.rownames[i];
-      break;
-    }
-  }
-  if (i == resultExp.rownames.length) {
-    return null;// cannot find gene
-  }
-  var tfaValues = [];
+  if (geneNames.length == 0)
+    return []; // cannot find gene
+
+  var allTfaValues = [];
+  var valueMin = Infinity;
+  var valueMax = -Infinity;
   if (fileTFA != null) {
-    var tfai = resultTFA.rownames.indexOf(name);
-    if (tfai != -1) {
-      for (var j = 0; j < resultTFA.numcols; j++) {
-        var idx = resultExp.colnames.indexOf(resultTFA.colnames[j]);
-        tfaValues.push({
-          value: resultTFA.values[tfai * resultTFA.numcols + j],
-          index: idx
-        });
-      }
+    geneNames.forEach(function(geneName, i) {
```

Comment:
I cannot find any more efficient way... I think traversing the geneNames and conditionNames, then check whether they are in TFA file or find their indices are unavoidable.